File: WondayWall/Services/GoogleAiService.cs
using System.IO;
using System.Net.Http;
using System.Text.Json;
using GenerativeAI;
using GenerativeAI.Exceptions;
using GenerativeAI.Types;
using Microsoft.Extensions.Logging;
using WondayWall.ComponentModel;
using WondayWall.Models;
using WondayWall.Utils;
using AppResources = WondayWall.Properties.Resources;

namespace WondayWall.Services;

public class GoogleAiService(
    AppConfigService configService,
    IHttpClientFactory httpClientFactory,
    ILogger<GoogleAiService> logger) : IAiService
{
    private const string GoogleAiApiKeyPageUrl = "https://aistudio.google.com/app/api-keys";
    private static string PaidTierRequiredMessage => AppResources.GoogleAiBillingError + GoogleAiApiKeyPageUrl;
    private const string TextModelName = "gemini-3-flash-preview";
    private const string ImageModelName = "gemini-3.1-flash-image-preview";

    private readonly HttpClient ogpHttpClient = httpClientFactory.CreateClient("WondayWall");
    private readonly HttpClient geminiHttpClient = httpClientFactory.CreateClient("Gemini");
    private static readonly JsonSerializerOptions JsonSerializerOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        WriteIndented = true,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    public async Task<GeneratedImageInfo> GenerateWallpaperAsync(
        PromptContext context,
        GoogleAiServiceTier serviceTier,
        CancellationToken ct = default)
    {
        var promptResult = await GeneratePromptAsync(context, serviceTier, ct).ConfigureAwait(false);
        var contextWithOgp = await FetchOgpImagesAsync(context, promptResult.SelectedNewsIds, ct).ConfigureAwait(false);
        return await GenerateImageFromPromptAsync(
            promptResult.ImagePrompt,
            contextWithOgp,
            promptResult.ServiceTier,
            ct).ConfigureAwait(false);
    }

    public async Task<PromptGenerationResult> GeneratePromptAsync(
        PromptContext context,
        GoogleAiServiceTier serviceTier,
        CancellationToken cancellationToken = default)
    {
        var config = configService.Current;

        if (string.IsNullOrWhiteSpace(config.GoogleAiApiKey))
            throw new InvalidOperationException("Google AI API key is not configured.");

        var result = await GeneratePromptSelectionWithFallbackAsync(
            context,
            serviceTier,
            config.GoogleAiApiKey,
            cancellationToken).ConfigureAwait(false);
        var selectedNews = ResolveSelectedNewsTopics(context, result.PromptSelection.SelectedNewsIds);

        return new PromptGenerationResult(
            ImagePrompt: result.PromptSelection.ImagePrompt,
            SelectedNewsTopics: selectedNews,
            SelectedNewsIds: result.PromptSelection.SelectedNewsIds,
            ServiceTier: result.ServiceTier);
    }

    public async Task<PromptContext> FetchOgpImagesAsync(
        PromptContext context,
        IReadOnlyList<string> selectedNewsIds,
        CancellationToken cancellationToken = default)
    {
        var selectedIds = selectedNewsIds.ToHashSet(StringComparer.Ordinal);
        var newsTopics = (context.NewsTopics ?? []).ToList();
        var targets = newsTopics
            .Select((topic, index) => (Topic: topic, Index: index))
            .Where(item => selectedIds.Contains(item.Topic.Id)
                && !string.IsNullOrWhiteSpace(item.Topic.OgpImageUrl))
            .Take(3)
            .ToList();

        var downloads = await Task.WhenAll(targets.Select(async item =>
        {
            try
            {
                using var response = await ogpHttpClient.GetAsync(
                    item.Topic.OgpImageUrl,
                    cancellationToken).ConfigureAwait(false);
                response.EnsureSuccessStatusCode();
                var mimeType = response.Content.Headers.ContentType?.MediaType ?? "image/jpeg";
                var data = await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
                return (Success: true, item.Index, Data: data, MimeType: mimeType);
            }
            catch (Exception ex) when (!cancellationToken.IsCancellationRequested)
            {
                logger.LogWarning(ex, "OGP画像のダウンロードに失敗しました [{ImgUrl}]", item.Topic.OgpImageUrl);
                return (Success: false, item.Index, Data: Array.Empty<byte>(), MimeType: "image/jpeg");
            }
        })).ConfigureAwait(false);

        foreach (var download in downloads.Where(item => item.Success))
        {
            newsTopics[download.Index] = newsTopics[download.Index] with
            {
                OgpImageData = download.Data,
                OgpImageMimeType = download.MimeType,
            };
        }

        return context with { NewsTopics = newsTopics };
    }

    public async Task<GeneratedImageInfo> GenerateImageFromPromptAsync(
        string imagePrompt,
        PromptContext context,
        GoogleAiServiceTier serviceTier,
        CancellationToken cancellationToken = default)
    {
        var config = configService.Current;
        if (string.IsNullOrWhiteSpace(config.GoogleAiApiKey))
            throw new InvalidOperationException("Google AI API key is not configured.");

        var imageRequest = BuildImageRequest(context, imagePrompt);
        return await GenerateImageWithFallbackAsync(
            context,
            imagePrompt,
            imageRequest,
            serviceTier,
            config.GoogleAiApiKey,
            cancellationToken).ConfigureAwait(false);
    }

    private async Task<PromptSelectionResult> GeneratePromptSelectionWithFallbackAsync(
        PromptContext context,
        GoogleAiServiceTier serviceTier,
        string apiKey,
        CancellationToken ct)
    {
        if (serviceTier != GoogleAiServiceTier.Flex)
            return await GeneratePromptSelectionAsync(context, GoogleAiServiceTier.Standard, apiKey, ct).ConfigureAwait(false);

        try
        {
            return await GeneratePromptSelectionAsync(context, GoogleAiServiceTier.Flex, apiKey, ct).ConfigureAwait(false);
        }
        catch (Exception ex) when (!ct.IsCancellationRequested)
        {
            logger.LogWarning(ex, "画像プロンプト生成の Flex 呼び出しが規定回数失敗したため Standard モードで再試行します。");
            return await GeneratePromptSelectionAsync(context, GoogleAiServiceTier.Standard, apiKey, ct).ConfigureAwait(false);
        }
    }

    private async Task<PromptSelectionResult> GeneratePromptSelectionAsync(
        PromptContext context,
        GoogleAiServiceTier serviceTier,
        string apiKey,
        CancellationToken ct)
    {
        // ステップ1: テキストモデルで詳細な画像プロンプトを生成（Google検索グラウンディングを有効化）
        var textModel = new GenerativeModelEx(apiKey, TextModelName, httpClient: geminiHttpClient, serviceTier: serviceTier);
        var contextPrompt = BuildTextModelPrompt(context);
        var promptRequest = new GenerateContentRequest();
        promptRequest.UseJsonMode<PromptSelectionResponse>(JsonSerializerOptions);
        promptRequest.AddText(contextPrompt);
        AddGoogleSearchTool(promptRequest);

        GenerateContentResponse promptResponse;
        try
        {
            promptResponse = await textModel.GenerateContentAsync(promptRequest, ct);
        }
        catch (ApiException ex) when (IsPaidTierRequiredError(ex))
        {
            throw new InvalidOperationException(PaidTierRequiredMessage, ex);
        }
        var promptSelection = promptResponse.ToObject<PromptSelectionResponse>(JsonSerializerOptions);

        if (promptSelection == null || string.IsNullOrWhiteSpace(promptSelection.ImagePrompt))
            throw new InvalidOperationException("Google AI returned an invalid structured prompt response.");

        promptSelection.SelectedNewsIds = (promptSelection.SelectedNewsIds ?? [])
            .Where(id => !string.IsNullOrWhiteSpace(id))
            .Select(id => id.Trim())
            .Distinct(StringComparer.Ordinal)
            .ToList();
        promptSelection.ImagePrompt = promptSelection.ImagePrompt.Trim();

        return new PromptSelectionResult(promptSelection, serviceTier);
    }

    private async Task<GeneratedImageInfo> GenerateImageWithFallbackAsync(
        PromptContext context,
        string imagePrompt,
        GenerateContentRequest imageRequest,
        GoogleAiServiceTier serviceTier,
        string apiKey,
        CancellationToken ct)
    {
        if (serviceTier != GoogleAiServiceTier.Flex)
        {
            return await GenerateImageAsync(context, imagePrompt, imageRequest, GoogleAiServiceTier.Standard, apiKey, ct).ConfigureAwait(false);
        }

        try
        {
            return await GenerateImageAsync(context, imagePrompt, imageRequest, GoogleAiServiceTier.Flex, apiKey, ct).ConfigureAwait(false);
        }
        catch (Exception ex) when (!ct.IsCancellationRequested)
        {
            logger.LogWarning(ex, "画像生成の Flex 呼び出しが規定回数失敗したため、生成済みプロンプトを使って Standard モードで再試行します。");
            return await GenerateImageAsync(context, imagePrompt, imageRequest, GoogleAiServiceTier.Standard, apiKey, ct).ConfigureAwait(false);
        }
    }

    private async Task<GeneratedImageInfo> GenerateImageAsync(
        PromptContext context,
        string imagePrompt,
        GenerateContentRequest imageRequest,
        GoogleAiServiceTier serviceTier,
        string apiKey,
        CancellationToken ct)
    {
        // ステップ2: 画像モデルでアスペクト比・サイズを指定して壁紙を生成
        var imageModel = new GenerativeModelEx(apiKey, ImageModelName, httpClient: geminiHttpClient, serviceTier: serviceTier);

        GenerateContentResponse response;
        try
        {
            response = await imageModel.GenerateContentAsync(imageRequest, ct).ConfigureAwait(false);
        }
        catch (ApiException ex) when (IsPaidTierRequiredError(ex))
        {
            throw new InvalidOperationException(PaidTierRequiredMessage, ex);
        }

        var imageData = ExtractImageBytes(response);

        if (imageData == null || imageData.Value.Bytes.Length == 0)
            throw new InvalidOperationException("No image data returned from Google AI.");

        var filePath = FileNameHelper.GetImageFilePath(PathUtility.WallpaperDirectory, extension: imageData.Value.Extension);
        await File.WriteAllBytesAsync(filePath, imageData.Value.Bytes, ct).ConfigureAwait(false);

        return new(filePath, DateTime.Now, imagePrompt, serviceTier, context);
    }

    private static GenerateContentRequest BuildImageRequest(PromptContext context, string imagePrompt)
    {
        var referenceImages = (context.NewsTopics ?? [])
            .Where(topic => topic.OgpImageData is { Length: > 0 })
            .Take(3)
            .ToList();
        var finalPrompt = referenceImages.Count > 0
            ? $$"""
              {{imagePrompt}}

              Reference images from the selected news topics are attached. Incorporate their visual themes, color palette, and subject matter into the wallpaper design.
              """
            : imagePrompt;

        var imageRequest = new GenerateContentRequest();

        imageRequest.AddText(finalPrompt);
        foreach (var referenceImage in referenceImages)
        {
            imageRequest.AddInlineData(
                Convert.ToBase64String(referenceImage.OgpImageData!),
                referenceImage.OgpImageMimeType ?? "image/jpeg");
        }

        var displayInfo = DisplayHelper.GetDisplayInfo();
        imageRequest.GenerationConfig = new GenerationConfig
        {
            ResponseModalities = [Modality.IMAGE],
            ImageConfig = new ImageConfig
            {
                AspectRatio = displayInfo.AspectRatio,
                ImageSize = displayInfo.ImageSize,
            }
        };
        AddGoogleSearchTool(imageRequest);

        return imageRequest;
    }

    private static List<NewsTopicItem> ResolveSelectedNewsTopics(
        PromptContext context,
        IReadOnlyList<string> selectedNewsIds)
    {
        var topicsById = (context.NewsTopics ?? [])
            .GroupBy(topic => topic.Id, StringComparer.Ordinal)
            .ToDictionary(group => group.Key, group => group.First(), StringComparer.Ordinal);

        return selectedNewsIds
            .Where(topicsById.ContainsKey)
            .Select(id => topicsById[id])
            .Select(topic => new NewsTopicItem(
                Title: topic.Title,
                Summary: topic.Summary,
                Url: topic.Url,
                PublishedAt: topic.PublishedAt,
                OgpImageUrl: topic.OgpImageUrl))
            .ToList();
    }

    private static void AddGoogleSearchTool(GenerateContentRequest request)
    {
        request.Tools ??= [];
        if (request.Tools.Any(static tool => tool.GoogleSearch != null))
            return;

        request.Tools.Add(new Tool
        {
            GoogleSearch = new GoogleSearchTool(),
        });
    }

    /// <summary>
    /// テキストモデルへ送るプロンプトを構築する。
    /// テキストモデルは候補コンテキストから採用要素を決め、画像生成用JSONを返す。
    /// </summary>
    private static string BuildTextModelPrompt(PromptContext context)
    {
        var parts = new List<string>
        {
            $$"""
            You are an expert desktop wallpaper image-generation prompt writer.
            You will be given calendar events, news topics, and optionally reference images from those news articles.
            You MUST aggressively use Google Search before writing the prompt.
            Research broadly and actively: run multiple targeted searches per topic (official sources, recent coverage,
            image references, and related background context), then cross-check recency and consistency.
            Prefer fresh, high-signal information and concrete visual details you can translate into imagery.
            Do not rely only on the user's short summaries when searchable context exists.
            Your task: review all candidate calendar events and news topics, decide which ones should materially influence
            the wallpaper, and then write a single detailed, creative English prompt for an image generation model
            ({{context.ImageSize}} resolution, {{context.AspectRatio}} aspect ratio) that creates a beautiful desktop wallpaper.

            The wallpaper should visually reflect the themes, mood, and atmosphere of the selected events and news.
            If reference images are supplied later, they will correspond only to selected news topics.
            Describe visual elements, style, mood, color palette, lighting, and composition in detail.
            No text, logos, or UI overlays. Wide landscape orientation unless aspect ratio specifies otherwise.

            For calendar events:
            - Only include POSITIVE events (celebrations, trips, parties, hobbies, achievements, social gatherings, etc.)
              in the visual design. Ignore NEGATIVE or NEUTRAL events (medical appointments, work deadlines,
              chores, administrative tasks, etc.), but do not let them suppress other event or news candidates.
            - Each event has a proximity tag indicating when it occurs. Use it to determine the visual weight:
              [today] or [tomorrow]: this event DOMINATES the entire image — make it the primary subject and theme,
                occupying nearly all visual elements.
              [in 2-3 days]: this event is a MAJOR visual theme, occupying 50–70% of the image's visual elements.
              [in 4-7 days]: this event is a MINOR accent or background element (15–30% of visual elements).
            - When multiple positive events are present, prioritize the ones happening sooner.
            - If the nearest event is NEGATIVE or NEUTRAL, ignore it and continue considering later positive events
              and news topics as potential primary themes.

            Return a response that matches the configured JSON schema.
            - imagePrompt must be a single detailed English prompt for the image model.
            - selectedNewsIds must contain only ids of news topics that materially influenced imagePrompt.
            - If no news topic is used, selectedNewsIds must be an empty array.
            - Do not output markdown fences or any extra explanation.
            """,
        };

        if ((context.CalendarEvents ?? []).Count > 0)
        {
            parts.Add(
                $$"""
                Calendar event candidates (JSON):
                {{JsonSerializer.Serialize(context.CalendarEvents, JsonSerializerOptions)}}
                """);
        }

        if ((context.NewsTopics ?? []).Count > 0)
        {
            parts.Add(
                $$"""
                News topic candidates (JSON):
                {{JsonSerializer.Serialize(context.NewsTopics, JsonSerializerOptions)}}
                """);
        }

        if (!string.IsNullOrWhiteSpace(context.AdditionalConstraints))
        {
            parts.Add($"Additional instructions: {context.AdditionalConstraints}");
        }

        return string.Join("\n\n", parts);
    }

    private static (byte[] Bytes, string Extension)? ExtractImageBytes(GenerateContentResponse response)
    {
        foreach (var candidate in response.Candidates ?? [])
        {
            foreach (var part in candidate.Content?.Parts ?? [])
            {
                if (part.InlineData?.MimeType?.StartsWith("image/", StringComparison.OrdinalIgnoreCase) == true &&
                    part.InlineData.Data != null)
                {
                    var extension = part.InlineData.MimeType["image/".Length..].Trim().ToLowerInvariant();
                    if (extension.Length == 0)
                        continue;

                    return (Convert.FromBase64String(part.InlineData.Data), extension);
                }
            }
        }
        return null;
    }

    private sealed class PromptSelectionResponse
    {
        public required string ImagePrompt { get; set; }

        public required List<string> SelectedNewsIds { get; set; }
    }

    private sealed record PromptSelectionResult(
        PromptSelectionResponse PromptSelection,
        GoogleAiServiceTier ServiceTier);

    private static bool IsPaidTierRequiredError(ApiException ex)
        => ex is { ErrorCode: 400, ErrorStatus: "FAILED_PRECONDITION" }
            or { ErrorCode: 429, ErrorStatus: "RESOURCE_EXHAUSTED" };
}

File: WondayWall.Android/app/src/main/kotlin/com/studiofreesia/wondaywall/services/GoogleAiService.kt
package com.studiofreesia.wondaywall.services

import com.google.genai.Client
import com.google.genai.types.Blob
import com.google.genai.types.Content
import com.google.genai.types.GenerateContentConfig
import com.google.genai.types.GoogleSearch
import com.google.genai.types.Part
import com.google.genai.types.Schema
import com.google.genai.types.Tool
import com.studiofreesia.wondaywall.models.GeneratedImageInfo
import com.studiofreesia.wondaywall.models.GoogleAiServiceTier
import com.studiofreesia.wondaywall.models.PromptCalendarEvent
import com.studiofreesia.wondaywall.models.PromptContext
import com.studiofreesia.wondaywall.models.PromptNewsTopic
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

// Google AI Gemini API を使った壁紙画像生成サービス（google-genai SDK 使用）
class GoogleAiService(
    private val appConfigService: AppConfigService,
    private val filesDir: File,
) {
    companion object {
        private const val TEXT_MODEL_NAME = "gemini-3-flash-preview"
        private const val IMAGE_MODEL_NAME = "gemini-3.1-flash-image-preview"
    }

    private val json = Json { ignoreUnknownKeys = true }

    // OGP 画像取得用 OkHttp クライアント（google-genai SDK に含まれないため独立して保持する）
    private val ogpHttpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    // 壁紙画像を生成してローカルに保存し、GeneratedImageInfo を返す
    // serviceTier: バックグラウンド生成時は Flex を指定（失敗時は Standard にフォールバック）
    suspend fun generateWallpaper(
        context: PromptContext,
        serviceTier: GoogleAiServiceTier = GoogleAiServiceTier.Standard,
    ): GeneratedImageInfo {
        val apiKey = appConfigService.getConfig().googleAiApiKey
        if (apiKey.isEmpty()) {
            throw IllegalStateException("Google AI API キーが設定されていません。")
        }

        val client = Client.builder().apiKey(apiKey).build()

        // ステップ1: テキストモデルで詳細な画像プロンプトを生成（Flex→Standard フォールバック付き）
        val promptSelection = generatePromptSelectionWithFallback(client, context, serviceTier)

        // ステップ2: 画像モデルで壁紙を生成（Flex→Standard フォールバック付き）
        val imageData = generateImageWithFallback(
            client = client,
            prompt = promptSelection.imagePrompt,
            context = context,
            selectedNewsIds = promptSelection.selectedNewsIds.toSet(),
            serviceTier = serviceTier,
        )

        // 画像をローカルに保存する
        val wallpapersDir = File(filesDir, "wallpapers").also { it.mkdirs() }
        val fileName = "wallpaper_${System.currentTimeMillis()}.${imageData.extension}"
        val file = File(wallpapersDir, fileName)
        file.writeBytes(imageData.bytes)

        return GeneratedImageInfo(
            filePath = file.absolutePath,
            imagePrompt = promptSelection.imagePrompt,
            selectedNewsIds = promptSelection.selectedNewsIds,
        )
    }

    // Flex → Standard フォールバック付きでプロンプトを生成する
    private suspend fun generatePromptSelectionWithFallback(
        client: Client,
        context: PromptContext,
        serviceTier: GoogleAiServiceTier,
    ): PromptSelectionResponse {
        if (serviceTier != GoogleAiServiceTier.Flex) {
            return generatePromptSelection(client, context, GoogleAiServiceTier.Standard)
        }
        return try {
            generatePromptSelection(client, context, GoogleAiServiceTier.Flex)
        } catch (e: IOException) {
            // Flex 容量不足（503/429）→ Standard で再試行
            if (isFlexUnavailableError(e)) {
                generatePromptSelection(client, context, GoogleAiServiceTier.Standard)
            } else throw e
        }
    }

    // テキストモデルで画像プロンプト候補を生成する
    private suspend fun generatePromptSelection(
        client: Client,
        context: PromptContext,
        serviceTier: GoogleAiServiceTier,
    ): PromptSelectionResponse = withContext(Dispatchers.IO) {
        val systemPrompt = buildTextModelPrompt(context)

        // レスポンススキーマ（imagePrompt + selectedNewsIds の JSON オブジェクト）
        val schema = Schema.builder()
            .type("OBJECT")
            .properties(
                mapOf(
                    "imagePrompt" to Schema.builder().type("STRING").build(),
                    "selectedNewsIds" to Schema.builder()
                        .type("ARRAY")
                        .items(Schema.builder().type("STRING").build())
                        .build(),
                )
            )
            .required(listOf("imagePrompt", "selectedNewsIds"))
            .build()

        val configBuilder = GenerateContentConfig.builder()
            .responseMimeType("application/json")
            .responseSchema(schema)
            .tools(listOf(Tool.builder().googleSearch(GoogleSearch.builder().build()).build()))
        if (serviceTier == GoogleAiServiceTier.Flex) {
            // Flex ティアは SDK の serviceTier フィールドで設定する
            configBuilder.serviceTier("flex")
        }
        val config = configBuilder.build()

        val content = Content.builder()
            .role("user")
            .parts(listOf(Part.builder().text(systemPrompt).build()))
            .build()

        val response = client.models.generateContent(TEXT_MODEL_NAME, content, config)
        val jsonText = response.text()
            ?: throw IOException("Google AI から有効なプロンプト選択レスポンスを取得できませんでした。")

        val selection = json.decodeFromString<PromptSelectionResponse>(jsonText)
        if (selection.imagePrompt.isEmpty()) {
            throw IOException("Google AI が有効な画像プロンプトを返しませんでした。")
        }
        selection
    }

    // Flex → Standard フォールバック付きで画像を生成する
    private suspend fun generateImageWithFallback(
        client: Client,
        prompt: String,
        context: PromptContext,
        selectedNewsIds: Set<String>,
        serviceTier: GoogleAiServiceTier,
    ): ImageData {
        if (serviceTier != GoogleAiServiceTier.Flex) {
            return generateImage(client, prompt, context, selectedNewsIds, GoogleAiServiceTier.Standard)
        }
        return try {
            generateImage(client, prompt, context, selectedNewsIds, GoogleAiServiceTier.Flex)
        } catch (e: IOException) {
            if (isFlexUnavailableError(e)) {
                generateImage(client, prompt, context, selectedNewsIds, GoogleAiServiceTier.Standard)
            } else throw e
        }
    }

    // 画像モデルで壁紙画像を生成する
    private suspend fun generateImage(
        client: Client,
        prompt: String,
        context: PromptContext,
        selectedNewsIds: Set<String>,
        serviceTier: GoogleAiServiceTier,
    ): ImageData = withContext(Dispatchers.IO) {
        // 選択済みニュースの OGP 画像 URL を収集する（最大3件）
        val ogpUrls = context.newsTopics
            .filter { selectedNewsIds.contains(it.id) }
            .mapNotNull { it.ogpImageUrl }
            .take(3)

        val imagePromptText = if (ogpUrls.isEmpty()) prompt else """
            $prompt

            Reference images from the selected news topics are attached. Incorporate their visual themes, color palette, and subject matter into the wallpaper design.
        """.trimIndent()

        val parts = mutableListOf<Part>()
        parts += Part.builder().text(imagePromptText).build()

        // 選択済みニュースの OGP 画像をダウンロードしてインラインデータとして添付する
        for (ogpUrl in ogpUrls) {
            try {
                val imgRequest = Request.Builder()
                    .url(ogpUrl)
                    .header("User-Agent", "Mozilla/5.0 (Linux; Android 14)")
                    .build()
                ogpHttpClient.newCall(imgRequest).execute().use { response ->
                    if (response.isSuccessful) {
                        val bytes = response.body?.bytes() ?: return@use
                        val mimeType = (response.header("Content-Type") ?: "image/jpeg")
                            .substringBefore(";").trim()
                        val blob = Blob.builder().mimeType(mimeType).data(bytes).build()
                        parts += Part.builder().inlineData(blob).build()
                    }
                }
            } catch (_: Exception) {
                // OGP 画像のダウンロード失敗は無視する
            }
        }

        val configBuilder = GenerateContentConfig.builder()
            .responseModalities(listOf("IMAGE"))
        if (serviceTier == GoogleAiServiceTier.Flex) {
            configBuilder.serviceTier("flex")
        }
        val config = configBuilder.build()

        val content = Content.builder()
            .role("user")
            .parts(parts)
            .build()

        val response = client.models.generateContent(IMAGE_MODEL_NAME, content, config)

        for (candidate in response.candidates() ?: emptyList()) {
            for (part in candidate.content()?.parts() ?: emptyList()) {
                val blob = part.inlineData() ?: continue
                if (blob.mimeType()?.startsWith("image/") != true) continue
                val bytes = blob.data() ?: continue
                val ext = blob.mimeType()!!.removePrefix("image/").let {
                    if (it.isEmpty()) "jpg" else it
                }
                return@withContext ImageData(bytes = bytes, extension = ext)
            }
        }
        throw IOException("Google AI から画像データを取得できませんでした。")
    }

    // テキストモデルに送るプロンプトを構築する
    private fun buildTextModelPrompt(context: PromptContext): String {
        val parts = mutableListOf<String>()

        parts += """
            You are an expert mobile wallpaper image-generation prompt writer.
            You will be given calendar events, news topics, and optionally reference images from those news articles.
            You MUST aggressively use Google Search before writing the prompt.
            Research broadly and actively: run multiple targeted searches per topic (official sources, recent coverage,
            image references, and related background context), then cross-check recency and consistency.
            Prefer fresh, high-signal information and concrete visual details you can translate into imagery.
            Do not rely only on the user's short summaries when searchable context exists.
            Your task: review all candidate calendar events and news topics, decide which ones should materially influence
            the wallpaper, and then write a single detailed, creative English prompt for an image generation model
            (${context.imageSize} resolution, ${context.aspectRatio} aspect ratio) that creates a beautiful Android wallpaper.

            The wallpaper should visually reflect the themes, mood, and atmosphere of the selected events and news.
            If reference images are supplied later, they will correspond only to selected news topics.
            Describe visual elements, style, mood, color palette, lighting, and composition in detail.
            No text, logos, or UI overlays. Portrait orientation.
            Place the most important visual elements in the center to avoid being hidden by status bar or navigation bar.

            For calendar events:
            - Only include POSITIVE events (celebrations, trips, parties, hobbies, achievements, social gatherings, etc.)
              in the visual design. Ignore NEGATIVE or NEUTRAL events (medical appointments, work deadlines,
              chores, administrative tasks, etc.), but do not let them suppress other event or news candidates.
            - Each event has a proximity tag indicating when it occurs. Use it to determine the visual weight:
              [today] or [tomorrow]: this event DOMINATES the entire image — make it the primary subject and theme,
                occupying nearly all visual elements.
              [in 2-3 days]: this event is a MAJOR visual theme, occupying 50–70% of the image's visual elements.
              [in 4-7 days]: this event is a MINOR accent or background element (15–30% of visual elements).
            - When multiple positive events are present, prioritize the ones happening sooner.
            - If the nearest event is NEGATIVE or NEUTRAL, ignore it and continue considering later positive events
              and news topics as potential primary themes.

            Return a response that matches the configured JSON schema.
            - imagePrompt must be a single detailed English prompt for the image model.
            - selectedNewsIds must contain only ids of news topics that materially influenced imagePrompt.
            - If no news topic is used, selectedNewsIds must be an empty array.
            - Do not output markdown fences or any extra explanation.
        """.trimIndent()

        if (context.calendarEvents.isNotEmpty()) {
            val eventsJson = json.encodeToString(
                ListSerializer(PromptCalendarEvent.serializer()),
                context.calendarEvents
            )
            parts += "Calendar event candidates (JSON):\n$eventsJson"
        }

        if (context.newsTopics.isNotEmpty()) {
            // OGP 画像 URL は後から添付するため、ここでは除外して送らない
            val topicsForPrompt = context.newsTopics.map {
                PromptNewsTopic(
                    id = it.id, title = it.title, summary = it.summary,
                    url = it.url, publishedAt = it.publishedAt, ogpImageUrl = null,
                )
            }
            val newsJson = json.encodeToString(
                ListSerializer(PromptNewsTopic.serializer()),
                topicsForPrompt
            )
            parts += "News topic candidates (JSON):\n$newsJson"
        }

        if (context.additionalConstraints.isNotEmpty()) {
            parts += "Additional instructions: ${context.additionalConstraints}"
        }

        return parts.joinToString("\n\n")
    }

    // Flex 容量不足を示すエラーか判定する（503 / 429）
    private fun isFlexUnavailableError(e: IOException): Boolean {
        val msg = e.message ?: return false
        return msg.contains("503") || msg.contains("429")
    }

    // 画像データの一時保持用
    private data class ImageData(val bytes: ByteArray, val extension: String)

    // テキストモデルのレスポンス（JSON スキーマに合わせた構造体）
    @Serializable
    private data class PromptSelectionResponse(
        val imagePrompt: String,
        val selectedNewsIds: List<String>,
    )
}
